cpp/src/parquet/page_index.cc

```diff
@@ -0,0 +1,174 @@
+                   static_cast<int>(src.size()));
+  decoder->Decode(dst, /*max_values=*/1);
+}
+
+template <>
+void Decode<ByteArrayType>(std::unique_ptr<ByteArrayDecoder>&, const std::string& src,
+                           ByteArray* dst) {
+  dst->len = static_cast<uint32_t>(src.size());
+  dst->ptr = reinterpret_cast<const uint8_t*>(src.c_str());
+}
+
+template <typename DType>
+class TypedColumnIndexImpl : public TypedColumnIndex<DType> {
+ public:
+  using T = typename DType::c_type;
+
+  TypedColumnIndexImpl(const ColumnDescriptor& descr,
+                       const format::ColumnIndex& column_index)
+      : column_index_(column_index) {
+    min_values_.reserve(column_index_.null_pages.size());
+    max_values_.reserve(column_index_.null_pages.size());
+    // Decode min and max values into a compact form (i.e. w/o null page)
+    auto plain_decoder = MakeTypedDecoder<DType>(Encoding::PLAIN, &descr);
+    T value;
+    for (size_t i = 0; i < column_index_.null_pages.size(); ++i) {
+      if (!column_index_.null_pages[i]) {
+        non_null_page_indices_.emplace_back(static_cast<int32_t>(i));
+        Decode<DType>(plain_decoder, column_index_.min_values[i], &value);
+        min_values_.emplace_back(value);
+        Decode<DType>(plain_decoder, column_index_.max_values[i], &value);
+        max_values_.emplace_back(value);
+      }
+    }
+  }
+
+  const std::vector<bool>& null_pages() const override {
+    return column_index_.null_pages;
+  }
+
+  const std::vector<std::string>& encoded_min_values() const override {
+    return column_index_.min_values;
+  }
+
+  const std::vector<std::string>& encoded_max_values() const override {
+    return column_index_.max_values;
+  }
+
+  BoundaryOrder boundary_order() const override {
+    return static_cast<BoundaryOrder>(static_cast<int>(column_index_.boundary_order));
+  }
+
+  bool has_null_counts() const override { return column_index_.__isset.null_counts; }
+
+  const std::vector<int64_t>& null_counts() const override {
+    return column_index_.null_counts;
+  }
+
+  const std::vector<T>& min_values() const override { return min_values_; }
+
+  const std::vector<T>& max_values() const override { return max_values_; }
+
+  const std::vector<int32_t> GetNonNullPageIndices() const override {
+    return non_null_page_indices_;
+  }
+
+ private:
+  /// Wrapped thrift column index.
+  const format::ColumnIndex column_index_;
+  /// Decoded typed min/max values. Null pages are set to std::nullopt.
+  std::vector<T> min_values_;
+  std::vector<T> max_values_;
+  /// A list of page indices for not-null pages.
+  std::vector<int32_t> non_null_page_indices_;
+};
+
+class OffsetIndexImpl : public OffsetIndex {
+ public:
+  explicit OffsetIndexImpl(const format::OffsetIndex& offset_index) {
+    page_locations_.reserve(offset_index.page_locations.size());
+    for (const auto& page_location : offset_index.page_locations) {
```

Comment:
I'm a little bit concerned about the copying around vectors?  How big due we typically expected indexes to be?

Reply:
The size depends on the number of pages in a column chunk. If we don't copy them and instead wrap a thrift OffsetIndex inside the implementation class, then it pays to get a vector of PageLocations (as we cannot expose thrift internals). 

I can refactor it to remove the vector accessor by providing only the per-page accessors. Then pimpl can be utilized to remove the virtual functions to reduce the cost for every access.

Reply:
i guess it was a general concern for both I just mentioned it here, @pitrou if you feel comfortable with it it seems OK to me.